File: broker/src/builder.rs
use std::future::Future;

use crate::{AppConfig, ConfigError, run_with_config, run_with_config_and_shutdown};

pub struct BrokerBuilder {
  config: AppConfig,
}

impl BrokerBuilder {
  pub fn new() -> Self {
    Self {
      config: AppConfig::default(),
    }
  }

  pub fn from_config(config: AppConfig) -> Self {
    Self { config }
  }

  pub fn from_env() -> Result<Self, ConfigError> {
    Ok(Self::from_config(AppConfig::load()?))
  }

  pub fn with_config(mut self, update: impl FnOnce(&mut AppConfig)) -> Self {
    update(&mut self.config);
    self
  }

  pub fn data_dir(mut self, value: impl Into<String>) -> Self {
    self.config.broker.data_dir = value.into();
    self
  }

  pub fn broker_bind_addr(mut self, value: impl Into<String>) -> Self {
    self.config.broker.bind_addr = value.into();
    self
  }

  pub fn admin_enabled(mut self, enabled: bool) -> Self {
    self.config.admin.enabled = enabled;
    self
  }

  pub fn admin_bind_addr(mut self, value: impl Into<String>) -> Self {
    self.config.admin.bind_addr = value.into();
    self
  }

  pub fn raft_enabled(mut self, enabled: bool) -> Self {
    self.config.raft.enabled = enabled;
    self
  }

  pub fn max_payload_bytes(mut self, value: usize) -> Self {
    self.config.broker.max_payload_bytes = value;
    self
  }

  pub fn max_fetch_records(mut self, value: usize) -> Self {
    self.config.broker.max_fetch_records = value;
    self
  }

  pub fn config(&self) -> &AppConfig {
    &self.config
  }

  pub fn config_mut(&mut self) -> &mut AppConfig {
    &mut self.config
  }

  pub fn into_config(self) -> AppConfig {
    self.config
  }

  pub async fn run(self) -> store::Result<()> {
    run_with_config(self.config).await
  }

  pub async fn run_with_shutdown<F>(self, shutdown: F) -> store::Result<()>
  where
    F: Future<Output = ()> + Send,
  {
    run_with_config_and_shutdown(self.config, shutdown).await
  }
}

impl Default for BrokerBuilder {
  fn default() -> Self {
    Self::new()
  }
}

File: broker/builder.go
package broker

import "context"

type Builder struct {
	cfg Config
}

func NewBuilder() *Builder {
	return &Builder{cfg: DefaultConfig()}
}

func BuilderFromConfig(cfg Config) *Builder {
	normalizeConfig(&cfg)
	return &Builder{cfg: cfg}
}

func BuilderFromEnv() (*Builder, error) {
	cfg, err := LoadConfig()
	if err != nil {
		return nil, err
	}
	return BuilderFromConfig(cfg), nil
}

func (b *Builder) WithConfig(update func(*Config)) *Builder {
	if update != nil {
		update(&b.cfg)
		normalizeConfig(&b.cfg)
	}
	return b
}

func (b *Builder) DataDir(value string) *Builder {
	b.cfg.Broker.DataDir = value
	return b
}

func (b *Builder) BrokerBindAddr(value string) *Builder {
	b.cfg.Broker.BindAddr = value
	return b
}

func (b *Builder) AdminEnabled(enabled bool) *Builder {
	b.cfg.Admin.Enabled = enabled
	return b
}

func (b *Builder) AdminBindAddr(value string) *Builder {
	b.cfg.Admin.BindAddr = value
	return b
}

func (b *Builder) Config() Config {
	return b.cfg
}

func (b *Builder) New(opts ...Option) (*Broker, error) {
	return New(b.cfg, opts...)
}

func (b *Builder) Run(ctx context.Context) error {
	return RunWithConfig(ctx, b.cfg)
}
